File: backend/app/services/auth_service.py
"""Authentication & session orchestration."""
from __future__ import annotations

from app.core.exceptions import ConflictError, ForbiddenError, UnauthorizedError
from app.core.security import (
    REFRESH,
    create_access_token,
    create_refresh_token,
    decode_token,
    hash_password,
    verify_password,
)
from app.models.base import utcnow
from app.models.enums import Role, UserStatus
from app.models.notification import Session
from app.models.user import RecruiterProfile, StudentProfile, User
from app.repositories.user_repo import (
    RecruiterProfileRepository,
    SessionRepository,
    StudentProfileRepository,
    UserRepository,
)
from app.schemas.auth import (
    AuthResponse,
    LoginRequest,
    RegisterRequest,
    TokenPair,
    UserPublic,
)


class AuthService:
    def __init__(
        self,
        users: UserRepository,
        students: StudentProfileRepository,
        recruiters: RecruiterProfileRepository,
        sessions: SessionRepository,
    ) -> None:
        self.users = users
        self.students = students
        self.recruiters = recruiters
        self.sessions = sessions

    @staticmethod
    def _public(user: User) -> UserPublic:
        return UserPublic(
            id=str(user.id),
            email=user.email,
            full_name=user.full_name,
            role=user.role,
            avatar_url=user.avatar_url,
            status=user.status.value,
            email_verified=user.email_verified,
            onboarding_completed=user.onboarding_completed,
            company_id=user.company_id,
        )

    async def _issue_tokens(self, user: User, *, ip: str | None, ua: str | None) -> TokenPair:
        access = create_access_token(
            user_id=str(user.id), role=user.role.value, email=user.email
        )
        refresh, jti, expires_at = create_refresh_token(user_id=str(user.id))
        await self.sessions.insert(
            Session(user_id=str(user.id), jti=jti, expires_at=expires_at, ip=ip, user_agent=ua)
        )
        return TokenPair(access_token=access, refresh_token=refresh)

    async def register(
        self, data: RegisterRequest, *, ip: str | None = None, ua: str | None = None
    ) -> AuthResponse:
        if await self.users.get_by_email(data.email):
            raise ConflictError("An account with this email already exists.")

        user = User(
            email=data.email.lower(),
            password_hash=hash_password(data.password),
            full_name=data.full_name,
            role=data.role,
            status=UserStatus.ACTIVE,
        )
        user = await self.users.insert(user)

        # create the role-specific profile
        if user.role == Role.STUDENT:
            await self.students.insert(StudentProfile(user_id=str(user.id)))
        elif user.role == Role.RECRUITER:
            await self.recruiters.insert(RecruiterProfile(user_id=str(user.id)))

        tokens = await self._issue_tokens(user, ip=ip, ua=ua)
        return AuthResponse(user=self._public(user), tokens=tokens)

    async def login(
        self, data: LoginRequest, *, ip: str | None = None, ua: str | None = None
    ) -> AuthResponse:
        user = await self.users.get_by_email(data.email)
        if not user or not verify_password(data.password, user.password_hash):
            raise UnauthorizedError("Invalid email or password.")
        if user.status in (UserStatus.SUSPENDED, UserStatus.BANNED):
            raise ForbiddenError("This account is not permitted to sign in.")

        await self.users.update(str(user.id), {"last_login_at": utcnow()})
        tokens = await self._issue_tokens(user, ip=ip, ua=ua)
        return AuthResponse(user=self._public(user), tokens=tokens)

    async def refresh(self, refresh_token: str) -> TokenPair:
        payload = decode_token(refresh_token, expected_type=REFRESH)
        jti = payload.get("jti", "")
        session = await self.sessions.get_by_jti(jti)
        if not session or session.revoked:
            raise UnauthorizedError("Refresh token is no longer valid.")

        user = await self.users.get(payload["sub"])
        if not user:
            raise UnauthorizedError("User not found.")

        # rotate: revoke old jti and issue a fresh pair
        await self.sessions.revoke(jti)
        return await self._issue_tokens(user, ip=session.ip, ua=session.user_agent)

    async def logout(self, refresh_token: str) -> None:
        try:
            payload = decode_token(refresh_token, expected_type=REFRESH)
            await self.sessions.revoke(payload.get("jti", ""))
        except UnauthorizedError:
            return  # already invalid; treat as logged out

File: backend/src/services/authService.js
"use strict";

/** Authentication & session orchestration. Mirrors app/services/auth_service.py. */

const { User, StudentProfile, RecruiterProfile, Session } = require("../models");
const { Role, UserStatus } = require("../models/enums");
const {
  REFRESH,
  createAccessToken,
  createRefreshToken,
  decodeToken,
  hashPassword,
  verifyPassword,
} = require("../utils/security");
const { ConflictError, ForbiddenError, UnauthorizedError } = require("../utils/errors");
const { publicUser } = require("../utils/serialize");
const { Cache } = require("../config/redis");

function tokenPair(accessToken, refreshToken) {
  return { access_token: accessToken, refresh_token: refreshToken, token_type: "bearer" };
}

async function issueTokens(user, { ip, ua }) {
  const access = createAccessToken({
    userId: String(user.id),
    role: user.role,
    email: user.email,
  });
  const { token: refresh, jti, expiresAt } = createRefreshToken({ userId: String(user.id) });
  await Session.create({
    user_id: String(user.id),
    jti,
    expires_at: expiresAt,
    ip: ip ?? null,
    user_agent: ua ?? null,
  });
  return tokenPair(access, refresh);
}

async function register(data, { ip = null, ua = null } = {}) {
  const existing = await User.findOne({ email: data.email.toLowerCase() });
  if (existing) throw new ConflictError("An account with this email already exists.");

  const user = await User.create({
    email: data.email.toLowerCase(),
    password_hash: await hashPassword(data.password),
    full_name: data.full_name,
    role: data.role,
    status: UserStatus.ACTIVE,
  });

  // create the role-specific profile
  if (user.role === Role.STUDENT) {
    await StudentProfile.create({ user_id: String(user.id) });
  } else if (user.role === Role.RECRUITER) {
    await RecruiterProfile.create({ user_id: String(user.id) });
  }

  const tokens = await issueTokens(user, { ip, ua });
  return { user: publicUser(user), tokens };
}

async function login(data, { ip = null, ua = null } = {}) {
  const user = await User.findOne({ email: data.email.toLowerCase() });
  if (!user || !(await verifyPassword(data.password, user.password_hash))) {
    throw new UnauthorizedError("Invalid email or password.");
  }
  if ([UserStatus.SUSPENDED, UserStatus.BANNED].includes(user.status)) {
    throw new ForbiddenError("This account is not permitted to sign in.");
  }

  user.last_login_at = new Date();
  await user.save();
  const tokens = await issueTokens(user, { ip, ua });
  return { user: publicUser(user), tokens };
}

async function refresh(refreshToken) {
  const payload = decodeToken(refreshToken, { expectedType: REFRESH });
  const jti = payload.jti || "";
  const session = await Session.findOne({ jti });
  if (!session || session.revoked) {
    throw new UnauthorizedError("Refresh token is no longer valid.");
  }

  const user = await User.findById(payload.sub);
  if (!user) throw new UnauthorizedError("User not found.");

  // rotate: revoke old jti and issue a fresh pair
  session.revoked = true;
  await session.save();
  return issueTokens(user, { ip: session.ip, ua: session.user_agent });
}

async function logout(refreshToken) {
  try {
    const payload = decodeToken(refreshToken, { expectedType: REFRESH });
    await Session.updateOne({ jti: payload.jti || "" }, { $set: { revoked: true } });
  } catch (err) {
    if (err instanceof UnauthorizedError) return; // already invalid; treat as logged out
    throw err;
  }
}

/** Update the current user's account fields (name, email, avatar). */
async function updateMe(user, data) {
  if (data.email && data.email.toLowerCase() !== user.email) {
    const email = data.email.toLowerCase();
    const existing = await User.findOne({ email });
    if (existing && String(existing.id) !== String(user.id)) {
      throw new ConflictError("That email is already in use.");
    }
    user.email = email;
    user.email_verified = false;
  }
  if (data.full_name !== undefined) user.full_name = data.full_name;
  if (data.avatar_url !== undefined) user.avatar_url = data.avatar_url;
  await user.save();
  await Cache.delete(`user:${String(user.id)}`);
  return publicUser(user);
}

/** Change the current user's password after verifying the old one. */
async function changePassword(user, currentPassword, newPassword) {
  const fresh = await User.findById(user.id);
  if (!fresh || !(await verifyPassword(currentPassword, fresh.password_hash))) {
    throw new UnauthorizedError("Current password is incorrect.");
  }
  fresh.password_hash = await hashPassword(newPassword);
  await fresh.save();
  await Cache.delete(`user:${String(user.id)}`);
}

/** Merge notification/privacy preferences into the user's settings. */
async function updateSettings(user, incoming) {
  const fresh = await User.findById(user.id);
  if (!fresh) throw new UnauthorizedError("User not found.");
  const current = fresh.settings ? fresh.settings.toObject() : {};
  const merged = {
    notifications: { ...(current.notifications || {}), ...(incoming.notifications || {}) },
    privacy: { ...(current.privacy || {}), ...(incoming.privacy || {}) },
  };
  fresh.settings = merged;
  await fresh.save();
  await Cache.delete(`user:${String(user.id)}`);
  return publicUser(fresh);
}

/** List the user's active (non-revoked, non-expired) sessions. */
async function listSessions(userId) {
  const now = new Date();
  const sessions = await Session.find({
    user_id: userId,
    revoked: false,
    expires_at: { $gt: now },
  })
    .sort({ created_at: -1 })
    .limit(50);
  return sessions.map((s) => ({
    id: String(s.id),
    user_agent: s.user_agent ?? null,
    ip: s.ip ?? null,
    created_at: s.created_at,
    expires_at: s.expires_at,
  }));
}

/** Revoke every session for the user (logout on all devices). */
async function logoutAll(userId) {
  await Session.updateMany(
    { user_id: userId, revoked: false },
    { $set: { revoked: true } }
  );
}

module.exports = {
  register,
  login,
  refresh,
  logout,
  issueTokens,
  tokenPair,
  updateMe,
  changePassword,
  updateSettings,
  listSessions,
  logoutAll,
};
